### PR TITLE
[RFC] api.c: fix deleting cgroup created on shared mnt point

### PR DESCRIPTION
**Pull request by kamalesh-babulal**
cgroup v1 allows mounting of two or more controllers on a single
mount point.  This fails currently when trying to delete the cgroup
recursively for all the mounted controllers because the cgroup is
already deleted by the first controller of the share mount point and
the controllers then on complaint about missing cgroup path.

Fix this issue by checking if the cgroup controllers share mount point
and ignore the error if the file doesn't exist, similar to ignoring if
the file exists during creation.

Fixes: https://github.com/libcgroup/libcgroup/issues/127

Signed-off-by: Kamalesh Babulal <kamalesh.babulal@oracle.com>

### CODE REVIEW

File: src/api.c [hunk 1]
```diff
@@ -2714,6 +2714,46 @@ static int cgroup_get_parent_name(struct cgroup *cgroup, char **parent)
...
+	int i, j;
+
+	pthread_rwlock_rdlock(&cg_mount_table_lock);
+
+	for (i = 0; cg_mount_table[i].name[0] != '\0'; i++) {
+		if (strncmp(cg_mount_table[i].name, controller,
+			    sizeof(cg_mount_table[i].name)) == 0) {
+
+			controller_mnt_point =
+				strdup(cg_mount_table[i].mount.path);
```
Comment (drakenclimber) [inline src/api.c:2734]: Do you need to `strdup` here?  What if you just grab the address of the mount path?
```
controller_mnt_point = &cg_mount_table[i].mount.path;
```

File: src/api.c [hunk 2]
```diff
@@ -2714,6 +2714,46 @@ static int cgroup_get_parent_name(struct cgroup *cgroup, char **parent)
...
+			if (controller_mnt_point == NULL)
+				goto error;
+
+			break;
+		}
+	}
+
+	for (j = 0; controller_mnt_point && j < i; j++) {
```
Comment (drakenclimber) [inline src/api.c:2742]: Thinking out loud here.... what if the shared mount point is later in `cg_mount_table` than `i`?  Will this code still find it?

File: src/api.c [hunk 3]
```diff
@@ -2714,6 +2714,46 @@ static int cgroup_get_parent_name(struct cgroup *cgroup, char **parent)
...
+			    sizeof(cg_mount_table[i].name)) == 0) {
+			ret = 1;
+			break;
+		}
+	}
+
+	free(controller_mnt_point);
+
+error:
+	pthread_rwlock_rdlock(&cg_mount_table_lock);
```
[the remark this answers is not in the document]
Comment (kamalesh-babulal) [reply src/api.c:2753]: Thanks for the review @drakenclimber :( that's right it should have been unlock.

File: src/api.c (hunk 1 above)
Comment (kamalesh-babulal) [reply src/api.c:2734]: Agree, that should be enough.

File: src/api.c [hunk 4]
```diff
@@ -2755,6 +2795,10 @@ static int cgroup_find_parent(struct cgroup *cgroup, char *controller,
 	cgroup_dbg("parent's name is %s\n", parent_path);
 
 	if (stat(child_path, &stat_child) < 0) {
+		if (is_cgrp_ctrl_mnt_shared(cgroup, controller)) {
```
Comment (drakenclimber) [inline src/api.c:2798]: More thinking out loud here... What if we kept track of all of the deleted paths?  Then we could compare the `child_path` with previously deleted paths.  If it's already been deleted, we can probably just `continue` and move on.

Thoughts?

File: src/api.c [hunk 5]
```diff
@@ -2714,6 +2714,46 @@ static int cgroup_get_parent_name(struct cgroup *cgroup, char **parent)
 	return ret;
 }
 
+/*
+ * Checks if the cgroup's controller shares the mount point with any other
+ * controller of the cgroup. Returns 1 if shared or 0.
+ */
+static int is_cgrp_ctrl_mnt_shared(struct cgroup *cgroup, char *controller)
```
Comment (drakenclimber) [inline src/api.c:2721]: I think if we go with this function, we really need to unit test it.  Should be fairly easy to do, and it should allow us to catch corner cases.

File: src/api.c (hunk 2 above)
Comment (kamalesh-babulal) [reply src/api.c:2742]: This function gets called when the `cgroup` is missing.  If the shared mount point is the after `i` and this is the first of the controllers sharing the mount point, we need to worry because something went wrong.

File: src/api.c (hunk 4 above)
Comment (kamalesh-babulal) [reply src/api.c:2798]: I did think about it and was biased towards the current implementation, due to the frequency of such use cases.  The check gets performed on error,  to confirm if the directory was deleted by the previous controller or if it's a real bug.  Another approach I was thinking of is adding `cg_mount_point->shared` flag, which gets set during `cgroup_init` and we would quickly check if it's a shared mount point, again in this approach we are assuming that a first of the controller sharing the mount point deleted it.

Comment (drakenclimber) [reply src/api.c:2798]: The `shared` flag is an interesting idea

Comment (kamalesh-babulal) [reply src/api.c:2798]: I have implemented the `shared_mnt` flag idea at https://github.com/kamalesh-babulal/libcgroup/commit/6e156aa1c1e195a60e811e1f4a541620356d76a1 please let me know if you like it better than the current approach.

**Comment by kamalesh-babulal** [comment]
**V2 Changes**
-  Implemented `shared_mnt` flag approach.

**Review by drakenclimber — APPROVED** [review]

**Comment by drakenclimber** [comment]
Pushed to main in commit 7f9134019910a6357ce0ab778caa50d957ef9ce3.  Thanks, @kamalesh-babulal
